[Hyperengage] Destination

zhadier39 opened the pull request:
<!-- Hello and thank you for contributing to Segment action-destinations! -->

<!-- Before opening your pull request, make sure you have added and ran unit
     tests and tested your change locally. Refer to our testing
     documentation for more information: https://github.com/segmentio/action-destinations/blob/main/docs/testing.md -->

<!-- If you have questions or issues please open a new issue or create a new discussion
     post in Github. -->

### Hyperengage Integration
Smart alerting for startup GTM teams, Hyperengage tracks thousands of data points to trigger smart alerts on hidden opportunities when the accounts are ready for upsell, or likely to churn. Our customers are sales, customer success, and Revops teams at B2B SaaS companies. 

### How our customers want to use Segment
Hyperengage customers may already have usage events in Segment. If they don't, we recommend Segment as a way to quickly connect product usage data to Hyperengage.

Once the integration is connected, Hyperengage uses the data, in combination with other data sources to provide actionable insights which would help Hyperengage customers to reach out to the right people at the right time.

![image](https://github.com/segmentio/action-destinations/assets/113626827/3134295e-be45-4ae6-9689-cb4d0eeba5ab)

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

joe-ayoub-segment commented:
hi @zhadier39 , thanks for raising this PR for a new Integration. 
Because this is for a new Integration I'd like to meet with you to get an understanding of the use-cases involved, events  and user identifiers to be sent to your platform. 

Would you be OK meeting with me to run through this? 

You can schedule a call with me here: 

http://calendly.com/joe_ayoub/integration-workshop 

Best regards,
Joe

saadhypng commented:
hey @joe-ayoub-segment , 

We have booked a time for tomorrow. 

Thanks

joe-ayoub-segment commented on packages/destination-actions/src/destinations/hyperengage/identify/index.ts:
```diff
@@ -0,0 +1,77 @@
+    created_at: {
+      type: 'string',
+      required: false,
+      description: 'The timestamp when the user was created',
```
Maybe explain the format for the timestamp. e.g. ISO 8601, and give an example.

joe-ayoub-segment reviewed (changes requested):
Hi @zhadier39 , 

Here's the complete PR review. 

Best regards,
Joe

joe-ayoub-segment commented:
hi @zhadier39 , just checking that you are OK with the feedback on this PR? 
Best regards,
Joe

zhadier39 commented:
Hi @joe-ayoub-segment ,

Thanks for reaching out. Yes the feedback is perfect, I've reviewed them and I'm currently addressing the comments. I'll push the updates shortly and will let you know if I have any questions.

Best regards,
@zhadier39

joe-ayoub-segment commented on packages/destination-actions/src/destinations/hyperengage/identify/index.ts:
```diff
@@ -30,6 +30,32 @@ const action: ActionDefinition<Settings, Payload> = {
+      label: 'First name',
+      default: {
+        '@if': {
+          exists: { '@path': '$.traits.firstName' },
+          then: { '@path': '$.traits.firstName' },
+          else: { '@path': '$.properties.firstName' }
```
```suggestion
          exists: { '@path': '$.traits.first_name' },
          then: { '@path': '$.traits.first_name' },
          else: { '@path': '$.properties.first_name' }
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/hyperengage/identify/index.ts:
```diff
@@ -30,6 +30,32 @@ const action: ActionDefinition<Settings, Payload> = {
+      label: 'Last name',
+      default: {
+        '@if': {
+          exists: { '@path': '$.traits.lastName' },
+          then: { '@path': '$.traits.lastName' },
+          else: { '@path': '$.properties.lastName' }
```
```suggestion
          exists: { '@path': '$.traits.last_name' },
          then: { '@path': '$.traits.last_name' },
          else: { '@path': '$.properties.last_name' }
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/hyperengage/identify/index.ts:
```diff
@@ -48,12 +74,19 @@ const action: ActionDefinition<Settings, Payload> = {
       description: 'The account id, to uniquely identify the account associated with the user',
       label: 'Account id',
-      default: { '@path': '$.groupId' }
+      default: {
+        '@if': {
+          exists: { '@path': '$.context.group_id' },
+          then: { '@path': '$.context.group_id' },
+          else: { '@path': '$.context.groupId' }
+        }
+      }
```
```suggestion
      default: {
        '@if': {
          exists: { '@path': '$.context.group_id' },
          then: { '@path': '$.context.group_id' },
          else: { '@path': '$.groupId' }
        }
      }
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/hyperengage/group/index.ts:
```diff
@@ -14,7 +14,13 @@ const action: ActionDefinition<Settings, Payload> = {
       description: 'The External ID of the account to send properties for',
       label: 'Account id',
-      default: { '@path': '$.groupId' }
+      default: {
+        '@if': {
+          exists: { '@path': '$.context.group_id' },
+          then: { '@path': '$.context.group_id' },
+          else: { '@path': '$.context.groupId' }
+        }
+      }
```
```suggestion
      default: {
        '@if': {
          exists: { '@path': '$.context.group_id' },
          then: { '@path': '$.context.group_id' },
          else: { '@path': '$.groupId' }
        }
      }
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/hyperengage/track/index.ts:
```diff
@@ -16,26 +16,32 @@ const action: ActionDefinition<Settings, Payload> = {
       description: 'The account id, to uniquely identify the account associated with the user',
       label: 'Account id',
-      default: { '@path': '$.groupId' }
+      default: {
+        '@if': {
+          exists: { '@path': '$.context.group_id' },
+          then: { '@path': '$.context.group_id' },
+          else: { '@path': '$.context.groupId' }
+        }
+      }
```
```suggestion
      default: {
        '@if': {
          exists: { '@path': '$.context.group_id' },
          then: { '@path': '$.context.group_id' },
          else: { '@path': '$.groupId' }
        }
      }
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/hyperengage/validateInput.ts:
```diff
@@ -107,5 +98,6 @@ export const validateInput = (
     delete properties.trial_end
     delete properties.website
   }
+  console.log(properties)
```
please remove the console.log

joe-ayoub-segment commented on packages/destination-actions/src/destinations/hyperengage/validateInput.ts:
```diff
@@ -67,11 +47,20 @@ export const validateInput = (
-      name: input?.name,
-      created_at: input?.created_at,
-      ...properties.traits
+    if (input?.name) {
+      properties.traits = {
+        email: input?.email,
+        name: input?.name,
+        created_at: input?.created_at,
+        ...properties.traits
+      }
+    } else if (input?.first_name || input?.last_name) {
+      properties.traits = {
+        email: input?.email,
+        name: `${input?.first_name} ${input?.last_name}}`,
+        created_at: input?.created_at,
+        ...properties.traits
+      }
```
Should we throw an error if input.name and input.first_name and input.last_name are all empty?

zhadier39 replied:
Yes, I think throwing an error for mandatory field missing should be the way to go here. Lemme see what I can do.

zhadier39 replied:
Hey @joe-ayoub-segment , I've resolved the default path issues and added an error incase none of the first_name and last_name or name fields are provided. Sorry for the inconvenience and thank you so much for your time

joe-ayoub-segment commented on packages/destination-actions/src/destinations/hyperengage/commonFields.ts:
```diff
@@ -0,0 +1,164 @@
+    label: 'Timezone',
+    default: {
+      '@if': {
+        exists: { '@path': 'context.timezone' },
```
```suggestion
        exists: { '@path': '$.context.timezone' },
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/hyperengage/validateInput.ts:
```diff
@@ -0,0 +1,102 @@
+import { Settings } from './generated-types'
```
There are linting errors because of the use of 'any'. You can either fix them or add the following line: 

/* eslint-disable @typescript-eslint/no-explicit-any */

Also, you shouldn't throw a regular Error object. Please throw a PayloadValidationError object. You can include it here: 

import { PayloadValidationError } from '@segment/actions-core'

joe-ayoub-segment commented on packages/destination-actions/src/destinations/hyperengage/validateInput.ts:
```diff
@@ -0,0 +1,102 @@
+        ...properties.traits
+      }
+    } else {
+      throw new Error('Either name, or first_name and last_name must be provided.')
```
```suggestion
      throw new PayloadValidationError('Either name, or first_name and last_name must be provided.')
```

joe-ayoub-segment reviewed (changes requested):
hi @zhadier39 , couple more things. 
Also the unit tests are failing. Can you make sure they pass please?

You can run them using this command: 

yarn jest --testPathPattern='./packages/destination-actions/src/destinations/hyperengage'

Also please make sure this command passes: 

yarn validate

zhadier39 commented:
> yarn validate

Hey @joe-ayoub-segment ,
When I run yarn validate, there's only a single error in someone else's action destination

![image](https://github.com/segmentio/action-destinations/assets/113626827/71bba51b-8f53-49c1-9763-b5d1bc62c75c)

I haven't meddled with any other action destinations as you can see in the PR.

joe-ayoub-segment commented:
> > yarn validate
> 
> Hey @joe-ayoub-segment , When I run yarn validate, there's only a single error in someone else's action destination
> 
> ![image](https://user-images.githubusercontent.com/113626827/272231574-71bba51b-8f53-49c1-9763-b5d1bc62c75c.png)
> 
> I haven't meddled with any other action destinations as you can see in the PR. What am I expected to do here?

Hi @zhadier39 if yarn validate passes that's great, nothing to do. 

Do the unit tests pass when you run this?

yarn jest --testPathPattern='./packages/destination-actions/src/destinations/hyperengage'

joe-ayoub-segment commented:
Looks like you probably fixed the tests. I'll run CI now.
